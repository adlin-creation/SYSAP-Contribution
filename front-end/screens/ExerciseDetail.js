import React from "react";
import {
  View,
  StyleSheet,
  Text,
  FlatList,
  TouchableOpacity,
  ScrollView,
  Image,
  Button,
} from "react-native";
import ExerciseService from "../services/ExerciceService";
import ReactPlayer from "react-player";

export default class ExerciseDetail extends React.Component {
  state = {
    exercise: null,
    steps: [],
};




  async componentDidMount() {

    const { idExercise } = this.props.route.params;
    try {
      await this.fetchData(idExercise);
    } catch (error) {
      console.error("Error fetching data:", error);
    }

  }


  async fetchData(idExercise) {
    try {
      const data = await ExerciseService.fetchExerciseById(idExercise);
      console.log("1");
      console.log(data);
      this.setState({ exercise: data });
    } catch (error) {
      console.error(
        "Une erreur s'est produite lors de la récupération des données.",
        error
      );
    }
  }


  render() {
    const { exercise, steps } = this.state;

    console.log("3");
    console.log(exercise);

    if (!exercise) return null;

    return (
      <View style={styles.mainContainer}>
          <Text style={styles.exTitle}>{exercise.ExerciseName}</Text>
          <Image
            source={{ uri: `../assets/images_test/${exercise.ExerciseImageURL}` }}
            style={styles.exerciseImage}
          />
          <Text>{exercise.ExerciseDescription}</Text>
          <Text>min : {exercise.ExerciseNumberRepetitionsMin}</Text>
          <Text>max : {exercise.ExerciseNumberRepetitionsMax}</Text>
          <View style={styles.playerContainer}>
              <ReactPlayer
                style={styles.player}
                url={`../assets/videos_test/${exercise.ExerciseExplanationVidURL}`}
                controls={true}
                onError={(e) => console.error("ReactPlayer error:", e)}
              />
            </View>
        <TouchableOpacity
          style={styles.retour}
          onPress={() => this.props.navigation.goBack()}
        >
          <Text style={{ color: "white" }}>Retour</Text>
        </TouchableOpacity>
      </View>
  );
  }
}


const styles = StyleSheet.create({
  container: {
    padding: 10,
  },
  mainContainer: {
    alignItems: "center",
    marginTop: 40,
  },
  mainTitle: {
    color: "brown",
    fontSize: 48,
    fontWeight: "bold",
    marginBottom: 20,
  },
  exTitle: {
    color: "black",
    fontSize: 36,
    fontWeight: "bold",
    marginBottom: 20,
    marginTop: 60,
    textDecorationLine:"underline",
  },
  exComplete: {
    margin: "3%",
    fontSize: 26,
  },
  exerciseImage: {
    width: 250,
    height: 250,
  },
  detailsButton: {
    color: "blue",
    textDecorationLine: "underline",
    fontSize: 16,
    fontWeight: "bold",
    textAlign: "center",
    padding: 10,
  },
  playerContainer: {
    marginTop: "3%",
    flex: 2,
    alignItems: "center",
    justifyContent: "center",
    marginBottom: "3%",
  },
  player: {
    width: "100%",
    aspectRatio: 16 / 9,
  },
  jouerButtonSeance: {
    backgroundColor: "green",
    padding: 40,
    alignItems: "center",
    width: 400,
    marginTop: 20,
    borderRadius: 10,
    position: "relative",
    overflow: "hidden",
    display: "flex",
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    paddingHorizontal: 40,
  },

  playTriangle: {
    width: 0,
    height: 0,
    borderLeftWidth: 0,
    borderRightWidth: 30,
    borderBottomWidth: 30,
    borderLeftColor: "transparent",
    borderRightColor: "transparent",
    borderBottomColor: "white",
  },
  jouerButtonText: {
    paddingRight: 30,
    color: "white",
    fontSize: 18,
    fontWeight: "bold",
  },
  retour: {
    backgroundColor: "#9C26B0",
    padding: 20,
    margin: 30,
    alignItems: "center",
    marginTop: 20,
    borderRadius: 20,
    paddingHorizontal: 200,
    color: "white",
  },
});
