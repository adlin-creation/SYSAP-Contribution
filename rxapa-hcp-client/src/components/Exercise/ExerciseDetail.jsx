import React from "react";
import { Row, Col, Input, Button, Form, Modal } from "antd";
import { CheckOutlined } from '@ant-design/icons'; // Ajout de l'importation
import { Controller, useForm } from "react-hook-form";
import axios from "axios";
import useToken from "../Authentication/useToken";
import Constants from "../Utils/Constants";
import "./Styles.css";

export default function ExerciseDetail({ exercise, refetchExercises }) {
  const { handleSubmit, control } = useForm();
  const { token } = useToken();

  
  const [isOpenModal, setIsOpenModal] = React.useState(false);
  const [isErrorMessage, setIsErrorMessage] = React.useState(false);
  const [message, setMessage] = React.useState("");

  function openModal(message, isError) {
    setMessage(message);
    setIsErrorMessage(isError);
    setIsOpenModal(true);
  }

  function closeModal() {
    setIsOpenModal(false);
    setMessage("");
    setIsErrorMessage(false);
  }

  const onSubmit = (data) => {
    axios
      .put(`${Constants.SERVER_URL}/update-exercise/${exercise.key}`, data, {
        headers: {
          Authorization: "Bearer " + token,
        },
      })
      .then((res) => {
        refetchExercises();
        openModal(res.data.message, false);
      })
      .catch((err) => {
        openModal(err.response.data.message, true);
      });
  };

  return (
    <Row justify="center" align="middle" style={{ minHeight: '50vh' }}>
      <Col span={12}>
        <Form layout="vertical" onFinish={handleSubmit(onSubmit)}>
          <Form.Item label="Nom d'exercice :">
            <Controller
              name="name"
              control={control}
              defaultValue={exercise.name}
              render={({ field: { onChange, value } }) => (
                <Input
                  onChange={onChange}
                  value={value}
                  placeholder="Entrez le nouveau nom d'exercice."
                  required
                />
              )}
            />
          </Form.Item>

          <Form.Item label="Description d'exercice">
            <Controller
              name="description"
              control={control}
              defaultValue={exercise.description}
              render={({ field: { onChange, value } }) => (
                <Input.TextArea
                  onChange={onChange}
                  value={value}
                  placeholder="Entrez la nouvelle description de l'exercice."
                  rows={4}
                  required
                />
              )}
            />
          </Form.Item>

          <Form.Item>
            <Button
              type="primary"
              htmlType="submit"
              icon={<CheckOutlined />}
            >
              Modifier
            </Button>
          </Form.Item>
        </Form>

        {isOpenModal && (
          <Modal
            title="Feedback"
            open={isOpenModal}
            onCancel={closeModal}
            footer={[
              <Button key="close" onClick={closeModal}>
                Fermer
              </Button>,
            ]}
          >
            <p style={{ color: isErrorMessage ? 'red' : 'green' }}>{message}</p>
          </Modal>
        )}
      </Col>
    </Row>
  );
}
