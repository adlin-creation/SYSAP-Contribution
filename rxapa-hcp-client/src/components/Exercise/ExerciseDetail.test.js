import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import axios from "axios";
import ExerciseDetail from "./ExerciseDetail";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { act } from "react-dom/test-utils";

jest.mock("axios");

const mockExercise = {
  key: "123",
  name: "Squat",
  description: "Un exercice pour les jambes",
  category: "Force",
  fitnessLevel: "Intermédiaire",
};

const queryClient = new QueryClient();

describe("ExerciseDetail Component", () => {
  it("should render exercise details", () => {
    render(
      <QueryClientProvider client={queryClient}>
        <ExerciseDetail exercise={mockExercise} refetchExercises={jest.fn()} />
      </QueryClientProvider>
    );

    expect(screen.getByDisplayValue("Squat")).toBeInTheDocument();
    expect(screen.getByDisplayValue("Un exercice pour les jambes")).toBeInTheDocument();
  });

  it("should enable edit mode", async () => {
    render(
      <QueryClientProvider client={queryClient}>
        <ExerciseDetail exercise={mockExercise} refetchExercises={jest.fn()} />
      </QueryClientProvider>
    );

    fireEvent.click(screen.getByRole("button", { name: /Modifier/i }));

    const nameInput = screen.getByDisplayValue("Squat");
    fireEvent.change(nameInput, { target: { value: "Squat Profond" } });

    expect(nameInput.value).toBe("Squat Profond");
  });

  it("should submit the form and update exercise", async () => {
    axios.put.mockResolvedValue({ data: { message: "Exercice mis à jour avec succès" } });

    render(
      <QueryClientProvider client={queryClient}>
        <ExerciseDetail exercise={mockExercise} refetchExercises={jest.fn()} />
      </QueryClientProvider>
    );

    fireEvent.click(screen.getByRole("button", { name: /Modifier/i }));

    const nameInput = screen.getByDisplayValue("Squat");
    fireEvent.change(nameInput, { target: { value: "Squat Avancé" } });

    fireEvent.click(screen.getByRole("button", { name: /Enregistrer/i }));

    await waitFor(() => {
      expect(axios.put).toHaveBeenCalledTimes(1);
      expect(axios.put).toHaveBeenCalledWith(expect.stringContaining("/update-exercise/123"), expect.any(Object), expect.any(Object));
    });

    expect(screen.getByText("Exercice mis à jour avec succès")).toBeInTheDocument();
  });
});