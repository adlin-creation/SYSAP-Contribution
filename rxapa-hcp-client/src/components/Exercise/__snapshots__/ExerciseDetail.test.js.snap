// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ExerciseDetail form snapshot 1`] = `
<div>
  Loading...
</div>
`;
